File: backend/tools/vendor_utils.py
"""Shared vendor normalization — single source of truth for all tools."""

from __future__ import annotations

from typing import Any

_VENDOR_MAP: dict[str, str] = {
    "google workspace": "GOOGLE_WORKSPACE",
    "google": "GOOGLE_WORKSPACE",
    "gwsuite": "GOOGLE_WORKSPACE",
    "g suite": "GOOGLE_WORKSPACE",
    "gsuite": "GOOGLE_WORKSPACE",
    "slack": "SLACK",
    "github": "GITHUB",
    "zoom": "ZOOM",
    "microsoft 365": "MICROSOFT_365",
    "microsoft365": "MICROSOFT_365",
    "m365": "MICROSOFT_365",
    "office 365": "MICROSOFT_365",
    "office365": "MICROSOFT_365",
    "dropbox": "DROPBOX",
    "salesforce": "SALESFORCE",
    "servicenow": "SERVICENOW",
    "jira": "JIRA",
    "confluence": "CONFLUENCE",
    "okta": "OKTA",
    "box": "BOX",
    "hubspot": "HUBSPOT",
    "zendesk": "ZENDESK",
    "asana": "ASANA",
    "notion": "NOTION",
    "articulate 360": "ARTICULATE_360",
    "articulate360": "ARTICULATE_360",
    "articulate": "ARTICULATE_360",
    "figma": "FIGMA",
}


def normalize_vendor(raw: str) -> str:
    """Coerce any vendor string to canonical uppercase enum value."""
    key = raw.strip().lower()
    return _VENDOR_MAP.get(key, raw.strip().upper().replace(" ", "_"))


def normalize_applications(raw: Any) -> list[dict[str, str]] | None:
    """Coerce GPT output into [{vendor: ...}] list. Returns None if invalid."""
    if isinstance(raw, str):
        raw = [raw]
    if isinstance(raw, dict):
        raw = [raw]
    if not isinstance(raw, list) or len(raw) == 0:
        return None
    result = []
    for item in raw:
        if isinstance(item, str):
            result.append({"vendor": normalize_vendor(item)})
        elif isinstance(item, dict) and "vendor" in item:
            result.append({"vendor": normalize_vendor(str(item["vendor"]))})
        else:
            return None
    return result


def normalize_users(raw: Any) -> list[dict[str, str]] | None:
    """Coerce GPT output into [{email: ...}] list. Returns None if invalid."""
    if isinstance(raw, str):
        raw = [raw]
    if isinstance(raw, dict):
        raw = [raw]
    if not isinstance(raw, list) or len(raw) == 0:
        return None
    result = []
    for item in raw:
        if isinstance(item, str):
            result.append({"email": item})
        elif isinstance(item, dict) and "email" in item:
            entry: dict[str, str] = {"email": str(item["email"])}
            if item.get("name"):
                entry["name"] = str(item["name"])
            result.append(entry)
        else:
            return None
    return result

File: agent-backend/src/workflow/vendorUtils.ts
/**
 * Vendor name normalisation — maps human-readable names to CloudFuze enum values.
 * Ported and expanded from the original Python vendor_utils.py.
 */

const VENDOR_MAP: Record<string, string> = {
  'slack': 'SLACK',
  'google workspace': 'GOOGLE_WORKSPACE',
  'google_workspace': 'GOOGLE_WORKSPACE',
  'gsuite': 'GOOGLE_WORKSPACE',
  'g suite': 'GOOGLE_WORKSPACE',
  'google': 'GOOGLE_WORKSPACE',
  'github': 'GITHUB',
  'github enterprise': 'GITHUB',
  'jira': 'JIRA',
  'jira software': 'JIRA',
  'confluence': 'CONFLUENCE',
  'zoom': 'ZOOM',
  'figma': 'FIGMA',
  'notion': 'NOTION',
  'salesforce': 'SALESFORCE',
  'hubspot': 'HUBSPOT',
  'okta': 'OKTA',
  'azure ad': 'AZURE_AD',
  'azure active directory': 'AZURE_AD',
  'microsoft 365': 'MICROSOFT_365',
  'm365': 'MICROSOFT_365',
  'office 365': 'MICROSOFT_365',
  'microsoft office': 'MICROSOFT_365',
  'dropbox': 'DROPBOX',
  'box': 'BOX',
  'docusign': 'DOCUSIGN',
  'articulate 360': 'ARTICULATE_360',
  'articulate360': 'ARTICULATE_360',
  'articulate': 'ARTICULATE_360',
  'zendesk': 'ZENDESK',
  'intercom': 'INTERCOM',
  'asana': 'ASANA',
  'monday': 'MONDAY',
  'monday.com': 'MONDAY',
  'linear': 'LINEAR',
  'trello': 'TRELLO',
  'miro': 'MIRO',
  'loom': 'LOOM',
  'grammarly': 'GRAMMARLY',
  'webex': 'WEBEX',
  'cisco webex': 'WEBEX',
  'teams': 'MICROSOFT_TEAMS',
  'microsoft teams': 'MICROSOFT_TEAMS',
  'aws': 'AWS',
  'amazon web services': 'AWS',
  'gcp': 'GOOGLE_CLOUD',
  'google cloud': 'GOOGLE_CLOUD',
  'azure': 'AZURE',
  'microsoft azure': 'AZURE',
  'stripe': 'STRIPE',
  'twilio': 'TWILIO',
  'sendgrid': 'SENDGRID',
  'datadog': 'DATADOG',
  'pagerduty': 'PAGERDUTY',
  'sentry': 'SENTRY',
  'splunk': 'SPLUNK',
  'tableau': 'TABLEAU',
  'looker': 'LOOKER',
  'dbt': 'DBT',
  'snowflake': 'SNOWFLAKE',
  'airtable': 'AIRTABLE',
  'clickup': 'CLICKUP',
  'basecamp': 'BASECAMP',
  'harvest': 'HARVEST',
  'bamboohr': 'BAMBOOHR',
  'workday': 'WORKDAY',
  'rippling': 'RIPPLING',
  'gusto': 'GUSTO',
};

export function normalizeVendor(name: string): string {
  if (!name) return '';
  const key = name.trim().toLowerCase();
  return VENDOR_MAP[key] ?? name.trim().toUpperCase().replace(/\s+/g, '_');
}

/**
 * Coerce various input formats into a list of vendor strings.
 * Accepts: "Slack", ["Slack", "Figma"], [{vendor: "Slack"}]
 */
export function normalizeVendors(input: any): string[] {
  if (!input) return [];

  if (typeof input === 'string') {
    return input.split(',').map((v) => normalizeVendor(v.trim())).filter(Boolean);
  }

  if (Array.isArray(input)) {
    return input.flatMap((item) => {
      if (typeof item === 'string') return [normalizeVendor(item)];
      if (item?.vendor) return [normalizeVendor(item.vendor)];
      return [];
    });
  }

  return [];
}

/**
 * Coerce user input into a list of email strings.
 * Accepts: "user@example.com", ["a@b.com", "c@d.com"], [{email: "a@b.com"}]
 */
export function normalizeUsers(input: any): string[] {
  if (!input) return [];

  if (typeof input === 'string') {
    return input.split(',').map((e) => e.trim()).filter(Boolean);
  }

  if (Array.isArray(input)) {
    return input.flatMap((item) => {
      if (typeof item === 'string') return [item.trim()];
      if (item?.email) return [item.email.trim()];
      return [];
    });
  }

  return [];
}
